File: packages/php/src/Resources/Wallet.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `wallet` resource: balances, the ledger, and top-ups (live keys only).
 * Accessed as `$client->wallet`.
 */
final class Wallet
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Balances, one per currency and environment.
     *
     * @return list<array<string, mixed>>
     */
    public function balances(): array
    {
        return Models::data($this->transport->request('GET', '/v1/wallet'));
    }

    /**
     * Ledger entries, newest first. Not cursor based: pass `before` = the
     * smallest `id` already seen; stop when fewer than `limit` come back.
     *
     * @param array{limit?: int, before?: int} $params
     * @return list<array<string, mixed>>
     */
    public function ledger(array $params = []): array
    {
        $query = Models::map($params, Models::LEDGER_QUERY, [], true);

        return Models::data($this->transport->request('GET', '/v1/wallet/ledger', $query));
    }

    /**
     * Start a top-up. Returns `{id, reference, authorization_url, access_code, amount, currency, status}`.
     *
     * @param array{amount: string, currency: string, channel: string, email: string} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function createTopup(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::TOPUP, ['amount', 'currency', 'channel']);

        return Models::object($this->transport->request('POST', '/v1/wallet/topups', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }
}

File: packages/java/src/main/java/io/opensms/Wallet.java
package io.opensms;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import io.opensms.models.LedgerEntry;
import io.opensms.models.Topup;
import io.opensms.models.WalletBalance;

import java.util.Collections;
import java.util.List;

/** Wallet balances, ledger and top-ups. Accessed as {@code client.wallet()}. */
public final class Wallet {

    private final ApiTransport t;

    Wallet(ApiTransport t) {
        this.t = t;
    }

    /** @return balances for the key's environment. */
    public List<WalletBalance> balances() {
        return data(t.get("/v1/wallet", null, ApiTransport.type(JsonNode.class)),
                ApiTransport.listOf(WalletBalance.class));
    }

    /** @return the newest ledger entries (server default page size). */
    public List<LedgerEntry> ledger() {
        return ledger(null, null);
    }

    /**
     * The ledger pages by id, not by cursor: pass the smallest {@code id} you have
     * seen as {@code before}; stop when fewer than {@code limit} entries come back.
     *
     * @param limit  1..200, or {@code null}.
     * @param before return entries with a smaller id, or {@code null}.
     * @return ledger entries, newest first.
     */
    public List<LedgerEntry> ledger(Integer limit, Long before) {
        return data(t.get("/v1/wallet/ledger", Query.of().add("limit", limit).add("before", before),
                ApiTransport.type(JsonNode.class)), ApiTransport.listOf(LedgerEntry.class));
    }

    /**
     * Start a payment to top up the wallet (live keys only).
     *
     * @param params amount, currency, channel and payer email.
     * @return the payment initialisation, including {@code authorizationUrl}.
     */
    public Topup createTopup(TopupParams params) {
        return createTopup(params, null);
    }

    /**
     * @param params  amount, currency, channel and payer email.
     * @param options per-call options.
     * @return the payment initialisation.
     */
    public Topup createTopup(TopupParams params, RequestOptions options) {
        return t.json("POST", "/v1/wallet/topups", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(Topup.class));
    }

    private static <T> List<T> data(JsonNode root, JavaType listType) {
        if (root == null || root.get("data") == null || root.get("data").isNull()) {
            return Collections.emptyList();
        }
        return Json.MAPPER.convertValue(root.get("data"), listType);
    }
}
